Improve template caching

rebkwok opened the pull request:
Fixes a bug in the last modified time used*, and improves the template reloading in development

- Cache the template using its content only
File content is cached based on both the content source file and the template file. [Here](https://github.com/opensafely-core/airlock/blob/8fc9f5cf85ef5bd445812f23adb9094c58918ebe/airlock/views/helpers.py#L95) we first check if the ETag (based on both file content and template) has changed; if it has, we then check the last modified time. However, we only use the content file's update time as the last_modified time, which means that if the template was updated after the content file, its changes wouldn't be registered.
  We don't want to use the template's last modified date, because on a redeploy, these dates will change, even if the template content hasn't changed, and we'll reload the file content unecessarily.  So we now use a hash of the template's source as the template cache key. Although this means we now need to load the template in order to get the template cache_id, django caches templates by default, so it should be cheap to do. 

- Reload the template if it has been modified
Django by default caches templates, even in development. The RendererTemplate classes are defined on the Renderer classes, so the templates themselves are never reloaded even if a template file changes. Django's autoreload watches for template changes and clears the template cache, but in order to pick up those changes we need to ensure the RendererTemplate is reloaded.

  Note that the template cache is in-memory, so starting and stopping the development server would clear it. However, if you've changed a template and reloaded the page in the browser, the eTag and LastModified headers will be updated, so even when you start/stop the server, the changes don't get picked up.

  We could turn off caching in development by using different settings, but I think it's preferable to keep development settings as close to production as possible.
  
* I _think_ this is also a bug in production; a workspace/request file would have an older modified time than an updated template file, so even if the cache ID has changed due to the template file changing, you'd still get the not-modified response.

inglesp commented on tests/unit/test_renderers.py:
```diff
@@ -28,15 +31,18 @@ def test_renderers_get_renderer_workspace(
+    # Set the update time on the content file to the current time to ensure
+    # it's the time used for the last_modified date (otherwise in CI the
+    # template will be more recent because its just been cloned)
+    time = datetime.now().timestamp()
```
What do you think about avoiding any dependency on the current time, and explicitly setting the times on the files we care about?

inglesp reviewed (commented):
I _think_ this change is ok, but I'd really appreciate a worked example of how the current behaviour breaks, and how this change fixes it.  This would help me understand what the potential implications are of this being a bug in production.

This could take the form of a full browser-based test case that fails without the change and passes with it, but that could be fiddly.

That's a quite a bit more work than I'd usually ask for in a PR, but the stakes are quite a bit higher.

I'd also appreciate reviews from others on @opensafely-core/team-rap.

rebkwok commented:
I had some other thoughts on this this evening, will do some rework tomorrow. Although this fixes the bug, it also means that we'll always have to refresh after a deploy (when the source code is all updated)

rebkwok commented on tests/unit/test_renderers.py:
```diff
@@ -171,3 +177,53 @@ def test_log_renderer_handles_ansi_colors(tmp_path):
+    response = renderer.get_response()
+    response.render()
+    assert response.rendered_content == "Foo123"
+    assert renderer.last_modified != "Thu, 27 Mar 2025 12:46:46 GMT"
```
This has gone now, but it was a leftover from a previous version of the code where it had set the time on the file to this specific datetime

rebkwok commented:
I've reworked this PR entirely, having realised that using the template timestamp isn't what we want (see commit comments and updated PR description).  It now caches the template based on content, and the first commit adds the failing test that demonstrated the current bug

evansd reviewed (approved):
This looks correct to me and, in terms of the production behaviour, is a simplification and improvement over what went before so I think we should merge.

Having to explicitly check for changes and force a reload is always a bit awkward, but given that this is only relevant in development I'm not worried. And it definitely _is_ a big improvement for the development experience.

Overall, I wonder how much benefit the caching is actually giving us? It may be that most of the win comes from improving the rendering experience as Richard's been working on and then we can do away with the caching logic altogether. But I'm happy this PR is a step forwards.

(And incidentally it was very easy to review thanks to the excellent comments and commit messages.)

evansd commented:
No, I agree that it's cheap and it's simpler to have it always on. It's just the fact that we have to do it is awkward.
